client/components/ContextMenu.jsx

```diff
@@ -0,0 +1,114 @@
+/** @jsx h */
+import {h} from 'preact';
+import cls from 'classnames';
+import ContextMenuItem from './ContextMenuItem';
+import PureComponent from '../lib/PureComponent';
+import {store} from '../store';
+import {elementIsOutside} from '../utils';
+
+import s from './ContextMenu.css';
+
+export default class ContextMenu extends PureComponent {
+  componentDidMount() {
+    this.boundingRect = this.node.getBoundingClientRect();
```

Comment:
Right now the contents of the menu are static, so this optimization makes sense. It'll be necessary to update this strategy if the menu is made dynamic enough that its sizing can change.